### PR TITLE
mpi: autogen errhan api

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

[skip warnings]

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
Rebased:

test:mpich/ch3/most
test:mpich/ch4/most

**Review by shintaro-iwasaki — COMMENTED** [review]
Thanks! Overall it looks good. Except for minor error checks, the main changes I would appreciate are:
1. Create another commit that implements `MPIR_Errhandler_free_impl()` before this change.
2. Create another commit that moves this `*errorclass > MPIR_Process.attrs.lastusedcode` check.

If we want to remove `ThreadSafeNoUpdate`, I want another commit that explains why we remove `ThreadSafeNoUpdate`.

File: src/mpi/errhan/errhan_impl.c [hunk 1]
```diff
@@ -123,60 +125,51 @@ void MPIR_File_get_errhandler_impl(MPI_File file, MPI_Errhandler * errhandler)
...
 
     /* We don't bother with the case where the errhandler is NULL;
      * in this case, the error handler was the original, MPI_ERRORS_ARE_FATAL,
      * which is builtin and can never be freed. */
     if (comm_ptr->errhandler != NULL) {
-        MPIR_Errhandler_release_ref(comm_ptr->errhandler, &in_use);
-        if (!in_use) {
-            MPIR_Errhandler_free(comm_ptr->errhandler);
-        }
+        MPIR_Errhandler_free_impl(comm_ptr->errhandler);
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/errhan/errhan_impl.c:141]: Separated commit https://github.com/pmodels/mpich/pull/4957/commits/acc703dce4628e6245e9764974c59235b35a8cb4 , which includes refactoring of `errhan_impl.c`.

File: src/mpi/errhan/dynerrutil.c [hunk 2]
```diff
@@ -289,7 +288,12 @@ int MPIR_Err_add_class(void)
      * a string.  */
     user_class_msgs[new_class] = 0;
 
-    return (new_class | ERROR_DYN_MASK);
+    new_class |= ERROR_DYN_MASK;
+    if (new_class > MPIR_Process.attrs.lastusedcode) {
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/errhan/dynerrutil.c:282]: Separated commit https://github.com/pmodels/mpich/pull/4957/commits/38aee9aca1c8fa7b29395d7a6a8b59fa9f0087a4 for all error utility impl functions.

File: src/binding/c/errhan_api.txt [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+MPI_Comm_set_errhandler:
+    .desc: Set the error handler for a communicator
+    .seealso: MPI_Comm_get_errhandler, MPI_Comm_call_errhandler
+    .extra: ignore_revoked_comm
+
+MPI_File_call_errhandler:
+    .desc: Call the error handler installed on a file
+
+MPI_File_create_errhandler:
+    .desc: Create a file error handler
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:139]: Fixed.

File: src/mpi/errhan/errhan_impl.c [hunk 4]
```diff
@@ -64,18 +63,21 @@ int MPIR_File_create_errhandler_impl(MPI_File_errhandler_function * file_errhand
...
+    *errhandler_ptr = errhan_ptr;
 
   fn_exit:
     return mpi_errno;
   fn_fail:
     goto fn_exit;
-}
+#else
+    *errhandler_ptr = NULL;
+    return MPI_SUCCESS;
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/errhan/errhan_impl.c]: Fixed.

File: src/binding/c/errhan_api.txt [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    by 'MPI_Comm_group' is no longer needed.
+*/
+
+MPI_Comm_set_errhandler:
+    .desc: Set the error handler for a communicator
+    .seealso: MPI_Comm_get_errhandler, MPI_Comm_call_errhandler
+    .extra: ignore_revoked_comm
+
+MPI_File_call_errhandler:
+    .desc: Call the error handler installed on a file
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:136]: Added `.docnotes: ThreadSafeNoUpdate` for all three `_call_errhandler` functions. Updated the python scripts to skip global critical section.

File: src/binding/c/errhan_api.txt [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+    According to the MPI-2 standard, it is erroneous to call 'MPI_Add_error_string'
+    for an error code or class with a value less than or equal
+    to 'MPI_ERR_LASTCODE'.  Thus, you cannot replace the predefined error messages
+    with this routine.
+*/
+
+MPIX_Delete_error_class:
+    errorclass: ERROR_CLASS, [value of the error class to be remoed]
+    .desc: Delete an MPI error class from the known classes
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:31]: `MPIR_ERRTEST_ARGNULL` checks whether a *pointer* argument is `NULL`. It shouldn't be applied to `errorclass`. The implementation (`MPIR_Delete_error_class_impl`) will throw error when the `errorclass` is wrong (missing in hash) so it is OK to skip the validation here.

File: src/binding/c/errhan_api.txt [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+MPIX_Delete_error_code:
+    errorcode: ERROR_CODE, [value of the error code to be remoed]
+    .desc: Delete an MPI error code
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:35]: Addressed in https://github.com/pmodels/mpich/pull/4957#discussion_r548619551

File: src/binding/c/errhan_api.txt [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+MPIX_Delete_error_string:
+    errorcode: ERROR_CODE, [value of the error code whose string is to be remoed]
+    .desc: Delete the error string associated with an MPI error code or class
+/*
+    Notes:
+    According to the MPI 4.1 standard, it is erroneous to call 'MPI_Delete_error_string'
+    for an error code or class with a value less than or equal
+    to 'MPI_ERR_LASTCODE'.  Thus, you cannot replace the predefined error messages
+    with this routine.
+*/
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:46]: Addressed in https://github.com/pmodels/mpich/pull/4957#discussion_r548619551

File: src/binding/c/errhan_api.txt [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+MPI_Errhandler_create:
+    comm_errhandler_fn: FUNCTION_SMALL, func_type=MPI_Comm_errhandler_function, [user defined error handling procedure]
+    errhandler: ERRHANDLER, direction=out
+    .desc: Creates an MPI-style errorhandler
+    .skip: ThreadSafe, initcheck
+    .replace: removed with MPI_Comm_create_errhandler
+    .seealso: MPI_Comm_create_errhandler, MPI_Errhandler_free
+
+MPI_Errhandler_free:
+    .desc: Frees an MPI-style errorhandler
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:81]: `errhandler` is an "inout" argument. I guess we should check for NULL before we access it with `*errhandler`. This applies to all `inout` handle argument. Let me add the rule to python script.

... Added.

File: src/binding/c/errhan_api.txt [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    comm: COMMUNICATOR, direction=lis:inout, param:in
+    errhandler: ERRHANDLER, [new error handler for communicator]
+    .desc: Sets the error handler for a communicator
+    .skip: initcheck
+    .replace: removed with MPI_Comm_set_errhandler
+    .seealso: MPI_Comm_set_errhandler, MPI_Errhandler_create, MPI_Comm_create_errhandler
+
+MPI_Comm_call_errhandler:
+    .desc: Call the error handler installed on a communicator
+    .extra: ignore_revoked_comm
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:100]: > 
> 
> Previously, this function is "ThreadSafeNoUpdate" and does not take a lock.
> Now this routine takes a lock (`MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX`) and becomes "ThreadSafe".
> Is this change intended and acceptable?
> 
> Even if there's a good reason to take a lock, I want to see a separate commit that explains this change.

Addressed in https://github.com/pmodels/mpich/pull/4957#discussion_r548602715

[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:100]: > 
> 
> Don't we need to check the following?
> 
> ```c
>             if (comm_ptr->errhandler) {
>                 MPIR_ERRTEST_ERRHANDLER(comm_ptr->errhandler->handle, mpi_errno);
>             }
> ```

The handle that is attached to comm (or win, or file) is always valid, or it won't get attached in the first place. So no need to check

File: src/binding/c/errhan_api.txt [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+MPI_Comm_create_errhandler:
+    .desc: Create a communicator error handler
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:103]: Added rule to the python script

File: src/binding/c/errhan_api.txt [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    return value from this routine is no longer needed.  This behavior is similar
+    to the other MPI routines for getting objects; for example, 'MPI_Comm_group'
+    requires that the user call 'MPI_Group_free' when the group returned
+    by 'MPI_Comm_group' is no longer needed.
+*/
+
+MPI_Comm_set_errhandler:
+    .desc: Set the error handler for a communicator
+    .seealso: MPI_Comm_get_errhandler, MPI_Comm_call_errhandler
+    .extra: ignore_revoked_comm
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:133]: Builtin errhandler will also pass `MPIR_Errhandler_valid_ptr`.

File: src/binding/c/errhan_api.txt [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    .desc: Call the error handler installed on a file
+
+MPI_File_create_errhandler:
+    .desc: Create a file error handler
+
+MPI_File_get_errhandler:
+    .desc: Get the error handler attached to a file
+
+MPI_File_set_errhandler:
+    .desc: Set the error handler for an MPI file
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:145]: > ThreadSafeNoUpdate

Added `.docnotes: ThreadSafeNoUpdate`

[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:145]: > ```c
> MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__,
>                                                  __LINE__, MPI_ERR_ARG, "**errhandnotfile", NULL);
> ```

Added custom error_check in `errhan_api.txt`.

File: src/binding/c/errhan_api.txt [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+MPI_Win_call_errhandler:
+    .desc: Call the error handler installed on a window
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:148]: Addressed in https://github.com/pmodels/mpich/pull/4957#discussion_r548602715

File: src/binding/c/errhan_api.txt [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+MPI_Win_create_errhandler:
+    .desc: Create an error handler for use with MPI window
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:151]: Addressed in https://github.com/pmodels/mpich/pull/4957#discussion_r548635241

File: src/binding/c/errhan_api.txt [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
+MPI_Win_get_errhandler:
+    .desc: Get the error handler for the MPI RMA window
+
+MPI_Win_set_errhandler:
+    .desc: Set window error handler
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/errhan_api.txt:157]: Added custom error_check in `errhan_api.txt`

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Review by shintaro-iwasaki — APPROVED** [review]
@hzhou.  Thanks a lot!  I approve this PR assuming that 1. you will fix the following and 2. pass the tests before merging this PR unless there's a special reason to leave them as `ThreadSafe`.

- Add `.ThreadSafeNoUpdate`
  - `MPI_Comm_get_errhandler()`
  - `MPI_Comm_set_errhandler()`
  - `MPI_File_get_errhandler()`
  - `MPI_Win_set_errhandler()`
  - (`MPI_Win_get_errhandler()`)

`MPI_Win_get_errhandler()` is not marked as `ThreadSafeNoUpdate`, but because `MPI_Win_set_errhandler()` and `MPI_Win_call_errhandler()` are `ThreadSafeNoUpdate`, this getter should also be `ThreadSafeNoUpdate`.

Sorry for missing them in the previous review.

**Comment by hzhou** [comment]
> 
> 
> @hzhou. Thanks a lot! I approve this PR assuming that 1. you will fix the following and 2. pass the tests before merging this PR unless there's a special reason to leave them as `ThreadSafe`.
> 
>     * Add `.ThreadSafeNoUpdate`
>       
>       * `MPI_Comm_get_errhandler()`
>       * `MPI_Comm_set_errhandler()`
>       * `MPI_File_get_errhandler()`
>       * `MPI_Win_set_errhandler()`
>       * (`MPI_Win_get_errhandler()`)
> 
> 
> `MPI_Win_get_errhandler()` is not marked as `ThreadSafeNoUpdate`, but because `MPI_Win_set_errhandler()` and `MPI_Win_call_errhandler()` are `ThreadSafeNoUpdate`, this getter should also be `ThreadSafeNoUpdate`.
> 
> Sorry for missing them in the previous review.

I added `.ThreadSafeNoUpdate` note to these functions just to be consistent with current man pages. However, on seconding reading `maint/doctext`, I am not sure the intended meaning of it. I For the mentioned `info` case, since user is allowed to set info outside MPI `Init/Finailze` window, we can't really protect the info with global lock. I guess that is where this note is originally intended for. I am not sure what is the object supposed to be in the errorhandler case. It seems to me that all there functions can and should be protected by the lock. The bottom line is, I don't think `ThreadSafeNoUpdate` implies not taking global lock. Therefore, I removed the commit that skips `global_cs` automatically based on `ThreadSafeNoUPdate`. If we need skip `global_cs`, just do that separately.

Let me know what do you think.

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by shintaro-iwasaki** [comment]
@hzhou I believe this PR should focus on translating the current mechanism as it is (except for very obvious issues) and let #4915 deal with a lock problem. In this sense, I would like to keep a global lock removed from `.ThreadSafeNoUpdate` functions.

**Comment by hzhou** [comment]
> 
> 
> @hzhou I believe this PR should focus on translating the current mechanism as it is

The key issue is "the current mechanism" is inconsistent:
```
$ grep ThreadSafeNoUpdate *_errhandler.c
comm_call_errhandler.c:.N ThreadSafeNoUpdate
comm_get_errhandler.c:.N ThreadSafeNoUpdate
comm_set_errhandler.c:.N ThreadSafeNoUpdate
file_call_errhandler.c:.N ThreadSafeNoUpdate
file_get_errhandler.c:.N ThreadSafeNoUpdate
file_set_errhandler.c:.N ThreadSafeNoUpdate
win_call_errhandler.c:.N ThreadSafeNoUpdate
win_set_errhandler.c:.N ThreadSafeNoUpdate

$ grep _ENTER.*MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX *_errhandler.c
comm_create_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
comm_get_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
comm_set_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
file_create_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
file_get_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
win_create_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
win_get_errhandler.c:    MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
```

>  In this sense, I would like to keep a global lock removed from `.ThreadSafeNoUpdate` functions.
> 

By your argument, it is wrong "to keep a global lock removed from `.ThreadSafeNoUpdate` functions", right?

**Comment by shintaro-iwasaki** [comment]
> "to keep a global lock removed from `.ThreadSafeNoUpdate` functions"

You're right. Sorry I misunderstood something. `.ThreadSafeNoUpdate` does not imply `.global_cs`. We need explicit `.global_cs` if necessary.

I am still afraid of significantly changing the behavior of functions in this PR (e.g., by adding a lock) since this PR is for auto-generation. Of course it is unclear whether the change is "changing the behavior" or "fixing inconsistency," but at least, `MPI_xxx_call_errhandler()` currently has no lock. If we can keep it removed for these functions (by adding `.global_cs`), I'm okay. If we will remove them, please create a different commit that explains this change.

**Comment by hzhou** [comment]
Added `.skip: global_cs` to `MPI_{Comm,Win,File}_call_errhandler`.

test:mpich/ch3/most
test:mpich/ch4/most

**Comment by shintaro-iwasaki** [comment]
@hzhou Thanks!

**Pull request merged**